tools/api_boost/api_boost.py

```diff
@@ -0,0 +1,166 @@
+#
+# api_boost.py --generate_api_type_database --generate_compilation_database \
+#   --build_api_booster
+
+import argparse
+import functools
+import json
+import os
+import multiprocessing as mp
+import pathlib
+import re
+import subprocess as sp
+
+# Temporary location of modified files.
+TMP_SWP_SUFFIX = '.tmp.swp'
+
+# Detect API #includes.
+API_INCLUDE_REGEX = re.compile('#include "(envoy/.*)/[^/]+\.pb\.(validate\.)?h"')
+
+
+# Update a C++ file to the latest API.
+def ApiBoostFile(llvm_include_path, debug_log, path):
+  print('Processing %s' % path)
+  # Run the booster
+  try:
+    result = sp.run([
+        './bazel-bin/external/envoy_dev/clang_tools/api_booster/api_booster',
+        '--extra-arg-before=-xc++',
+        '--extra-arg=-isystem%s' % llvm_include_path, '--extra-arg=-Wno-undefined-internal', path
+    ],
+                    capture_output=True,
+                    check=True)
+  except sp.CalledProcessError as e:
+    print('api_booster failure for %s: %s %s' % (path, e, e.stderr.decode('utf-8')))
+    raise
+  if debug_log:
+    print(result.stderr.decode('utf-8'))
+
+  # Consume stdout containing the list of inferred API headers. We don't have
+  # rewrite capabilities yet in the API booster, so we rewrite here in Python
+  # below.
+  inferred_api_includes = sorted(set(result.stdout.decode('utf-8').splitlines()))
+
+  # We just dump the inferred API header includes at the start of the #includes
+  # in the file and remove all the present API header includes. This does not
+  # match Envoy style; we rely on later invocations of fix_format.sh to take
+  # care of this alignment.
+  output_lines = []
+  include_lines = ['#include "%s"' % f for f in inferred_api_includes]
+  input_text = pathlib.Path(path).read_text()
+  for line in input_text.splitlines():
+    if include_lines and line.startswith('#include'):
+      output_lines.extend(include_lines)
+      include_lines = None
+    # Exclude API includes, except for a special case related to v2alpha
+    # ext_authz; this is needed to include the service descriptor in the build
+    # and is a hack that will go away when we remove v2.
+    if re.match(API_INCLUDE_REGEX, line) and 'envoy/service/auth/v2alpha' not in line:
+      continue
+    output_lines.append(line)
+
+  # Write to temporary file. We can't overwrite in place as we're executing
+  # concurrently with other ApiBoostFile() invocations that might need the file
+  # we're writing to.
+  pathlib.Path(path + TMP_SWP_SUFFIX).write_text('\n'.join(output_lines) + '\n')
+
+
+# Replace the original file with the temporary file created by ApiBoostFile()
+# for a given path.
+def SwapTmpFile(path):
+  pathlib.Path(path + TMP_SWP_SUFFIX).rename(path)
+
+
+# Update the Envoy source tree the latest API.
+def ApiBoostTree(args):
+  # Optional setup of state. We need the API type database, compilation database
+  # and api_booster tool in place before we can start boosting.
+  if args.generate_api_type_database:
```

Comment:
this is no op because `@envoy_dev//clang_tools/api_booster` depends on the type_db in bazel level.